mesonbuild/interpreterbase.py

```diff
@@ -424,18 +424,30 @@ class KwargInfo(T.Generic[_T]):
         itself contain mutable types, typed_kwargs will copy the default
     :param since: Meson version in which this argument has been added. defaults to None
     :param deprecated: Meson version in which this argument has been deprecated. defaults to None
+    :param validator: A callable that does additional validation. This is mainly
+        intended for cases where a string is expected, but only a few specific
+        values are accepted. Must return None if the input is valid, or a
+        message if the input is invalid
+    :param convertor: A callable that converts the raw input value into a
+        different type. This is intended for cases such as the meson DSL using a
+        string, but the implementation using an Enum. This should not do
+        validation, just converstion.
     """
 
     def __init__(self, name: str, types: T.Union[T.Type[_T], T.Tuple[T.Type[_T], ...], ContainerTypeInfo],
                  required: bool = False, listify: bool = False, default: T.Optional[_T] = None,
-                 since: T.Optional[str] = None, deprecated: T.Optional[str] = None):
+                 since: T.Optional[str] = None, deprecated: T.Optional[str] = None,
+                 validator: T.Optional[T.Callable[[_T], T.Optional[str]]] = None,
+                 convertor: T.Optional[T.Callable[[_T], T.Any]] = None):
```

Comment:
Maybe use a second `T.TypeVar` instead of `T.Any` or does this not work?

Reply:
I wanted to use object there, but it doesn't fit the type annotations. I guess maybe we could do this with an overload? let met try that.

Reply:
Okay, I couldn't get object working, but it seems happy with `TYPE_nvar`, so I guess we can got with that for now.